lib/timers.js

```diff
@@ -626,9 +636,6 @@ exports.clearImmediate = function(immediate) {
 
   immediate._onImmediate = undefined;
 
-  L.remove(immediate);
-
-  if (L.isEmpty(immediateQueue)) {
-    process._needImmediateCallback = false;
-  }
+  // leave queued, much faster overall to skip it later in processImmediate
```

Comment:
I would expect actual applications make very few calls to clearImmediate
so you're most likely right.  I was focusing on the effect of deferring the remove a bit.

The output of test from benchmark/timers/immediate.js
With the remove as it existed:
`timers/immediate.js thousands=2000 type=clear: 1388.82567`
Modified to not remove (ie, the git diff shown above for lines 629-633 vs 639):
`timers/immediate.js thousands=2000 type=clear: 3536.98189`
The test:
```
function clear(N) {
  bench.start();
  function cb(a1) {
    if (a1 === 2)
      bench.end(N / 1e3);
  }
  for (var i = 0; i < N; i++) {
    clearImmediate(setImmediate(cb, 1));
  }
  setImmediate(cb, 2);
}
```
This test creates 2m cleared immediates and shows 254% speedup.
My other test that loops 10k gets 302%.